File: ios/Reguerta/Reguerta/Data/Firestore/FirestoreReviewerEnvironmentRouter.swift
import FirebaseFirestore
import Foundation

final class FirestoreReviewerEnvironmentRouter: @unchecked Sendable, ReviewerEnvironmentRouter {
    private let db: Firestore

    init(db: Firestore = Firestore.firestore()) {
        self.db = db
    }

    func applyRouting(for principal: AuthPrincipal) async {
        guard ReguertaRuntimeEnvironment.baseFirestoreEnvironment == .production else {
            ReguertaRuntimeEnvironment.resetToBaseEnvironment()
            return
        }
        let policy = await loadReviewerPolicy()
        let normalizedEmail = principal.email.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        let isAllowlisted = policy.emails.contains(normalizedEmail) || policy.uids.contains(principal.uid)
        let reviewerRoutingEnabled = MemberPermissionMatrix.reviewerCapabilities.contains(.routeProductionReviewerToDevelop)
        ReguertaRuntimeEnvironment.applySessionEnvironment((isAllowlisted && reviewerRoutingEnabled) ? .develop : .production)
    }

    func resetToBaseEnvironment() {
        ReguertaRuntimeEnvironment.resetToBaseEnvironment()
    }

    private func loadReviewerPolicy() async -> ReviewerRoutingPolicy {
        let candidates = [
            "production/plus-collections/config/global",
            "production/collections/config/global",
            "production/config/global",
        ]

        for path in candidates {
            do {
                let snapshot = try await db.document(path).getDocument()
                guard snapshot.exists, let payload = snapshot.data() else { continue }
                return ReviewerRoutingPolicy.from(payload: payload)
            } catch {
                continue
            }
        }
        return .empty
    }
}

private struct ReviewerRoutingPolicy {
    let emails: Set<String>
    let uids: Set<String>

    static let empty = ReviewerRoutingPolicy(emails: [], uids: [])

    static func from(payload: [String: Any]) -> ReviewerRoutingPolicy {
        let rootEmails = payload.extractStringSet(
            keys: ["reviewerAllowlistEmails", "reviewerAllowlist", "reviewerEmails"]
        )
        let rootUids = payload.extractStringSet(
            keys: ["reviewerAllowlistUids", "reviewerUids"]
        )
        let nestedAllowlist = payload["reviewerAllowlist"] as? [String: Any]
        let nestedEmails = nestedAllowlist?.extractStringSet(keys: ["emails", "allowlistedEmails"]) ?? []
        let nestedUids = nestedAllowlist?.extractStringSet(keys: ["uids", "allowlistedUids"]) ?? []
        return ReviewerRoutingPolicy(
            emails: Set((rootEmails + nestedEmails).map { $0.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() }),
            uids: Set((rootUids + nestedUids).map { $0.trimmingCharacters(in: .whitespacesAndNewlines) })
        )
    }
}

private extension Dictionary where Key == String, Value == Any {
    func extractStringSet(keys: [String]) -> [String] {
        guard !keys.isEmpty else { return [] }
        var values = [String]()
        for key in keys {
            let value = self[key]
            switch value {
            case let string as String:
                let normalized = string.trimmingCharacters(in: .whitespacesAndNewlines)
                if !normalized.isEmpty {
                    values.append(normalized)
                }
            case let array as [Any]:
                for item in array {
                    guard let string = item as? String else { continue }
                    let normalized = string.trimmingCharacters(in: .whitespacesAndNewlines)
                    if !normalized.isEmpty {
                        values.append(normalized)
                    }
                }
            default:
                break
            }
        }
        return values
    }
}

File: android/Reguerta/app/src/main/java/com/reguerta/user/data/firestore/FirestoreReviewerEnvironmentRouter.kt
package com.reguerta.user.data.firestore

import com.google.android.gms.tasks.Tasks
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.FirebaseFirestore
import com.reguerta.user.domain.access.AccessCapability
import com.reguerta.user.domain.access.AuthPrincipal
import com.reguerta.user.domain.access.MemberPermissionMatrix
import com.reguerta.user.presentation.access.ReviewerEnvironmentRouter
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

class FirestoreReviewerEnvironmentRouter(
    private val firestore: FirebaseFirestore,
) : ReviewerEnvironmentRouter {
    override suspend fun applyRoutingFor(principal: AuthPrincipal) {
        if (ReguertaRuntimeEnvironment.baseFirestoreEnvironment != ReguertaFirestoreEnvironment.PRODUCTION) {
            ReguertaRuntimeEnvironment.resetToBaseEnvironment()
            return
        }
        val reviewerPolicy = loadReviewerPolicy()
        val normalizedEmail = principal.email.trim().lowercase()
        val isAllowlisted = reviewerPolicy.emails.contains(normalizedEmail) ||
            reviewerPolicy.uids.contains(principal.uid)
        val reviewerRoutingEnabled = MemberPermissionMatrix.reviewerCapabilities.contains(
            AccessCapability.ROUTE_PRODUCTION_REVIEWER_TO_DEVELOP,
        )
        val targetEnvironment = if (isAllowlisted && reviewerRoutingEnabled) {
            ReguertaFirestoreEnvironment.DEVELOP
        } else {
            ReguertaFirestoreEnvironment.PRODUCTION
        }
        ReguertaRuntimeEnvironment.applySessionEnvironment(targetEnvironment)
    }

    override fun resetToBaseEnvironment() {
        ReguertaRuntimeEnvironment.resetToBaseEnvironment()
    }

    private suspend fun loadReviewerPolicy(): ReviewerRoutingPolicy = withContext(Dispatchers.IO) {
        val candidates = listOf(
            "production/plus-collections/config/global",
            "production/collections/config/global",
            "production/config/global",
        )
        val snapshot = candidates.firstNotNullOfOrNull { path ->
            runCatching { Tasks.await(firestore.document(path).get()) }
                .getOrNull()
                ?.takeIf { it.exists() }
        } ?: return@withContext ReviewerRoutingPolicy.EMPTY
        snapshot.toReviewerRoutingPolicy()
    }
}

private data class ReviewerRoutingPolicy(
    val emails: Set<String>,
    val uids: Set<String>,
) {
    companion object {
        val EMPTY = ReviewerRoutingPolicy(emptySet(), emptySet())
    }
}

private fun DocumentSnapshot.toReviewerRoutingPolicy(): ReviewerRoutingPolicy {
    val payload = data.orEmpty()
    val rootEmails = payload.extractStringSet(
        "reviewerAllowlistEmails",
        "reviewerAllowlist",
        "reviewerEmails",
    )
    val rootUids = payload.extractStringSet(
        "reviewerAllowlistUids",
        "reviewerUids",
    )
    val nestedAllowlist = payload["reviewerAllowlist"] as? Map<*, *>
    val nestedEmails = nestedAllowlist.extractStringSet(
        "emails",
        "allowlistedEmails",
    )
    val nestedUids = nestedAllowlist.extractStringSet(
        "uids",
        "allowlistedUids",
    )
    return ReviewerRoutingPolicy(
        emails = (rootEmails + nestedEmails).mapTo(linkedSetOf()) { it.trim().lowercase() },
        uids = (rootUids + nestedUids).mapTo(linkedSetOf()) { it.trim() },
    )
}

private fun Map<*, *>?.extractStringSet(vararg keys: String): Set<String> {
    if (this == null || keys.isEmpty()) return emptySet()
    val values = linkedSetOf<String>()
    keys.forEach { key ->
        val value = this[key]
        when (value) {
            is String -> {
                val normalized = value.trim()
                if (normalized.isNotEmpty()) {
                    values += normalized
                }
            }

            is List<*> -> {
                value.forEach { item ->
                    val normalized = (item as? String)?.trim()
                    if (!normalized.isNullOrEmpty()) {
                        values += normalized
                    }
                }
            }
        }
    }
    return values
}
